src/transforms/aws_ec2_metadata.rs

```diff
@@ -0,0 +1,596 @@
+        let fields = self
+            .fields
+            .clone()
+            .map(|v| v.into_iter().map(Atom::from).collect())
+            .unwrap_or_else(|| DEFAULT_FIELD_WHITELIST.clone());
+
+        exec.spawn_std(
+            async move {
+                let mut client = MetadataClient::new(host, keys, write, refresh_interval, fields);
+
+                // Keep retrying the client if there are errors.
+                loop {
+                    if let Err(error) = client.fetch_metadata().await {
+                        error!(message = "Unable to refresh metadata.", %error);
+
+                        Delay::new(Instant::now() + Duration::from_secs(5))
+                            .compat()
+                            .await
+                            .unwrap();
+                    } else {
+                        break;
+                    }
+                }
+            }
+                // TODO: Once #1338 is done we can fetch the current span
+                .instrument(info_span!("aws_ec2_metadata: worker")),
+        );
+
+        Ok(Box::new(Ec2MetadataTransform { state: read }))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn transform_type(&self) -> &'static str {
+        "aws_ec2_metadata"
+    }
+}
+
+impl Transform for Ec2MetadataTransform {
+    fn transform(&mut self, mut event: Event) -> Option<Event> {
+        let log = event.as_mut_log();
+
+        self.state.for_each(|k, v| {
+            if let Some(value) = v.get(0) {
+                log.insert_explicit(k.clone(), value.clone().into());
+            }
+        });
+
+        Some(event)
+    }
+}
+
+struct MetadataClient {
+    client: Client<HttpConnector, Body>,
+    host: Uri,
+    token: Option<(Bytes, Instant)>,
+    keys: Keys,
+    state: WriteHandle,
+    refresh_interval: Duration,
+    fields: HashSet<Atom>,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct DynamicIdentityDocument {
+    account_id: String,
+    architecture: String,
+    image_id: String,
+    instance_id: String,
+    instance_type: String,
+    private_ip: String,
+    region: String,
+    version: String,
+}
+
+impl MetadataClient {
+    pub fn new(
+        host: Uri,
+        keys: Keys,
+        state: WriteHandle,
+        refresh_interval: Duration,
+        fields: Vec<Atom>,
+    ) -> Self {
+        Self {
+            client: Client::new(),
+            host,
+            token: None,
+            keys,
+            state,
+            refresh_interval,
+            fields: fields.into_iter().collect(),
+        }
+    }
+
+    async fn fetch_metadata(&mut self) -> Result<(), crate::Error> {
+        loop {
+            self.fetch_all().await?;
+
+            // Make changes viewable to the transform. This may block if
+            // readers are still reading.
+            self.state.refresh();
```

Comment:
It feels weird for this call to be separated from where we actually write to the map. It seems like we should either move this to the bottom of `fetch_all` (which might be better named `update_metadata` or something) or change the structure to something like

```rust
let data = self.fetch_all().await?
self.state.write(data);
self.state.refresh();
```

The way it's currently written, `fetch_all` has a weird implicit contract that it does side effects but requires you do to more work for them to become visible.